### ISSUE TITLE
Suggestion: Product Embedding Creation for e-commerce domain.

### ISSUE THREAD

**Issue by karndeepsingh** [clip-italian/clip-italian#11]
Hi,
I need a suggestion to build Product Embeddings for the E-commerce domain. I am working on a product embedding creation and the requirement is to create a single embedding for each product using the Product's Image, description, Price, and brand as Features. If you can provide some direction,it would help to process and understand the things properly.

Thanks.

**Comment by vinid** [comment]
Hello!

the most relevant thing to this repo and to the topic you are interested in is probably [FashionCLIP](https://arxiv.org/pdf/2204.03972.pdf).

These are some things on e-commerce data representation I did in the last couple of years: [1](https://arxiv.org/pdf/2003.07160.pdf), [2](https://arxiv.org/abs/2007.14906), [3](https://arxiv.org/pdf/2104.09423.pdf), [4](https://aclanthology.org/2021.naacl-industry.20/), [5](https://aclanthology.org/2021.ecnlp-1.1/). You can probably look at the related work section to find other papers.

You might also want to take a look at this [blog post](https://medium.com/nvidia-merlin/winning-the-sigir-ecommerce-challenge-on-session-based-recommendation-with-transformers-v2-793f6fac2994) by the NVIDIA team.

**Comment by karndeepsingh** [comment]
> 



> Hello!
> 
> the most relevant thing to this repo and to the topic you are interested in is probably [FashionCLIP](https://arxiv.org/pdf/2204.03972.pdf).
> 
> These are some things on e-commerce data representation I did in the last couple of years: [1](https://arxiv.org/pdf/2003.07160.pdf), [2](https://arxiv.org/abs/2007.14906), [3](https://arxiv.org/pdf/2104.09423.pdf), [4](https://aclanthology.org/2021.naacl-industry.20/), [5](https://aclanthology.org/2021.ecnlp-1.1/). You can probably look at the related work section to find other papers.
> 
> You might also want to take a look at this [blog post](https://medium.com/nvidia-merlin/winning-the-sigir-ecommerce-challenge-on-session-based-recommendation-with-transformers-v2-793f6fac2994) by the NVIDIA team.

Thanks @vinid for sharing the resource. I will definitely go through it. I have the following few more questions on FashionCLIP:
1. How can we add Product Price as the feature with the Product's image and Description in the FashionCLIP model. I have added Brand and Type of product in the description of each product but the Price of the product is also something I need to consider while forming embedding with image, and description. Please, let me know how I can add price information to the FashionCLIP.

Thanks.
Waiting for your suggestion.

**Comment by vinid** [comment]
Assuming we are talking about fine-tuning another CLIP model, you could append the price to the description, albeit I am not sure how well it is going to work. It mostly depends on how much data you have access to.

**Comment by karndeepsingh** [comment]
> Assuming we are talking about fine-tuning another CLIP model, you could append the price to the description, albeit I am not sure how well it is going to work. It mostly depends on how much data you have access to.

I will add price information in the description and will check how it performs. Will let you know.

Thanks again.

**Comment by karndeepsingh** [comment]
Hi @vinid,
I just wanted to understand. How can i can evaluate the embeddings generated by the CLIP model?

I have two embeddings outputted by CLIP model that is image and text embeddings for a Product and My objective is to get a single embedding for a product.  Can you suggest me How I can create a single embedding using these two image and text embedding outputted by the CLIP model?

And please also advise on how to evaluate the embeddings generated by the CLIP model. I went through FashionCLIP model, HITS@K metric has been used. Can you help me with some resource calculate HITS@k for my dataset?

Thanks.
Waiting for your suggestion

**Comment by vinid** [comment]
```I have two embeddings outputted by CLIP model that is image and text embeddings for a Product and My objective is to get a single embedding for a product. Can you suggest me How I can create a single embedding using these two image and text embedding outputted by the CLIP model?```

You could either average or pass the vectors to another network to train. It mostly depends on the use case.

```And please also advise on how to evaluate the embeddings generated by the CLIP model. I went through FashionCLIP model, HITS@K metric has been used. Can you help me with some resource calculate HITS@k for my dataset?```

A basic measure you can implement to evaluate CLIP is by doing image retrieval from captions and should be something like this:

assume you have `I`, a set of image embeddings and `T` a set of description embeddings, where `I[i]` is described by `T[i]` (e.g., a pic of a cat and the description "a picture of a cat"). This means that images and captions are aligned.

Now, let's consider `T[10]`. This is image number 10. 

We compute the similarity between vector `T[10]` and all the elements in `I` and build a list of similarities. Now, get the argmax of all this list of similarities (i.e., the index of the maximum value in the similarity vector).

If the `argmax == 10` (our image index) the model has found the correct image. 

You can generalize this to Top-K arguments and to all the texts in `T`.

**Comment by karndeepsingh** [comment]
> `I have two embeddings outputted by CLIP model that is image and text embeddings for a Product and My objective is to get a single embedding for a product. Can you suggest me How I can create a single embedding using these two image and text embedding outputted by the CLIP model?`
> 
> You could either average or pass the vectors to another network to train. It mostly depends on the use case.
> 
> `And please also advise on how to evaluate the embeddings generated by the CLIP model. I went through FashionCLIP model, HITS@K metric has been used. Can you help me with some resource calculate HITS@k for my dataset?`
> 
> A basic measure you can implement to evaluate CLIP is by doing image retrieval from captions and should be something like this:
> 
> assume you have `I`, a set of image embeddings and `T` a set of description embeddings, where `I[i]` is described by `T[i]` (e.g., a pic of a cat and the description "a picture of a cat"). This means that images and captions are aligned.
> 
> Now, let's consider `T[10]`. This is image number 10.
> 
> We compute the similarity between vector `T[10]` and all the elements in `I` and build a list of similarities. Now, get the argmax of all this list of similarities (i.e., the index of the maximum value in the similarity vector).
> 
> If the `argmax == 10` (our image index) the model has found the correct image.
> 
> You can generalize this to Top-K arguments and to all the texts in `T`.

@vinid Thanks again for answering. 
Following are the implementation code for finding matches using TOP-K,
```
def find_matches(model, image_embeddings, query, image_filenames, n=9):
    tokenizer = RobertaTokenizer.from_pretrained(CFG.text_tokenizer)
    encoded_query = tokenizer(query,padding=True,truncation=True)
    batch = {
        key: torch.tensor(values).to(CFG.device)
        for key, values in encoded_query.items()
    }
    with torch.no_grad():
        text_features = model.text_encoder(
            input_ids=batch["input_ids"], attention_mask=batch["attention_mask"]
        )
        text_embeddings = model.text_projection(text_features)
    
    image_embeddings_n = F.normalize(image_embeddings, p=2, dim=-1)
    text_embeddings_n = F.normalize(text_embeddings, p=2, dim=-1)
    dot_similarity = text_embeddings_n @ image_embeddings_n.T
    # print(dot_similarity.shape)
    #output_values = F.softmax(values, dim=-1)
    values, indices = torch.topk(dot_similarity.squeeze(0), n * 5)
    
    matches = [image_filenames[idx] for idx in indices[::5].to("cpu")]

    

    return matches
```

Just want to measure HITS@K or Mean Recipocal Rank to this output. Any suggestion what I need to add or remove?

**Comment by vinid** [comment]
i have skimmed it quickly and it looks good to me

**Comment by karndeepsingh** [comment]
> i have skimmed it quickly and it looks good to me

@vinid Thanks for your answers. 
 I would like to ask a few more questions:

1. I have images of the fashion product and its attributes like gender, color, type, size etc. how I can prepare text using these attributes so that I can use that text to pair with the image? And later while inferencing, I can extract all these attributes of the given new image. If you can give me some idea on preparing the text of the attribute so that I can train the CLIP model for better context training?

Please help me to understand the text preparation step as I have classes of a particular product as its attributes like color, type, size, gender, etc, and how I can prepare a better description out of these attributes and then while inferencing I can extract these attributes from the new image.

Thanks.

**Comment by vinid** [comment]
Hello

it depends on what you have. If you have no descriptions at all it might be better to not use CLIP in this context. In theory, you could concatenate elements to sentences "a **blue** **large** **shirt** for **men**" where in bold you find the attributes, not sure how well this can generalize in practice.

You can then extract those attributes using zero-shot classification.

However, isn't it better to just train a classifier in this context since you have images and attributes?

**Comment by karndeepsingh** [comment]
> Hello
> 
> it depends on what you have. If you have no descriptions at all it might be better to not use CLIP in this context. In theory, you could concatenate elements to sentences "a **blue** **large** **shirt** for **men**" where in bold you find the attributes, not sure how well this can generalize in practice.
> 
> You can then extract those attributes using zero-shot classification.
> 
> However, isn't it better to just train a classifier in this context since you have images and attributes?

Thanks again @vinid for replying. 
I was planning to solve this problem using Multi-label classification  as mentioned. But Since, we have lot of attributes and distribution is different and it would be difficult to maintain the different models trained on different attributes as a multi-label classification problem hence, I was thinking to go in an unsupervised way so that I can just train a single model that captures all the attributes and then later with zero-shot classification we can extract attributes.
